design/proposals/module-linking/Explainer.md

```diff
@@ -838,34 +842,60 @@ binaries in all the same places where core module binaries can be loaded today,
 branching on the `layer` field in the binary format to determine the kind of
 module to decode.
 
-As with the JS API, the main question is how ESM-integration should deal with
-single-level imports, instance imports and module imports. Going through these
-one at a time:
+As with the JS API, the main question for ESM-integration is how to deal with
+all imports having a single string as well as the new instance and module
+import/export types. Going through these:
+
+For adapter module imports of module type, we need a fundamentally new way to
+request that the ESM loader parse or decode a module without *also*
+instantiating that module. Recognizing this same need from JavaScript, there is
+already a TC39 proposal called [Import Reflection] that adds the ability to
+write, in JavaScript:
+```js
+import Foo from "./foo.wasm" as "wasm-module";
+assert(Foo instanceof WebAssembly.Module);
+```
+With this extension to JavaScript and the ESM loader, an adapter module import
+of module type can be treated like an `as "wasm-module"` import by
+ESM-integration.
```

Comment:
> Note that strictly speaking, this exception isn't needed per the current spec since the reflection module would still be a default export, and thus follow the logic of the last case in this text.

Ahhh, so, to check my understanding: is the idea that, when you write `as "wasm-module"`, the import *does* produce a "Module Record", however, that Module Record isn't a normal Module Record (with live exports, etc), but a rather specialized Module Record that has a `default` export which contains a [`module`](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-module) produced by [`module_decode`](https://webassembly.github.io/spec/core/appendix/embedding.html#mathrm-module-decode-xref-syntax-values-syntax-byte-mathit-byte-ast-xref-syntax-modules-syntax-module-mathit-module-xref-appendix-embedding-embed-error-mathit-error)?  (I was assuming you didn't get a Module Record at all.)

> What if an adapter module exports a module type as a named export, is that use case possible? This might be useful in bundling workflows?

Yes, that's possible: it should produce a `WebAssembly.Module` as the exported value (instead of the usual `WebAssembly.`{`Function`,`Table`,`Memory`,...}).  In this PR, this is very briefly covered by the end of the JS API section and then referenced at the end of the ESM-integration section, but let me know if you think that should be expanded.

> What if you want to both load the same module as an instance and as its module reflection? Not sure this has any real use though.

Ooh, iiinteresting: so iiuc the problem is that the "no duplicate import/export names" rule would prevent you from having two imports with the exact same Module Specifier (one with instance type and one with module type).  Does the Module Specifier logic have any normalization logic so that you could have two literally-different import strings but they both turn into the same normalized string?  Otherwise, you could always use two totally-different import strings and use an import map to map one to the other...

Reply:
> Ahhh, so, to check my understanding: is the idea that, when you write as "wasm-module", the import does produce a "Module Record", however, that Module Record isn't a normal Module Record (with live exports, etc), but a rather specialized Module Record that has a default export which contains a module produced by module_decode? (I was assuming you didn't get a Module Record at all.)

Since it hasn't been specified yet it's actually not clear, but even from the import side, it would currently be a default import (so sugar for `import { default as module } from './x.wasm' as "wasm-module"`). So that the equivalence remains on the importer side, unless the reflection spec were to entirely ban named imports and treat it entirely as a value type import exception. I don't think any change is necessary here, was just noting it might fall out using the normal logic without exception, definitely worth waiting on further spec clarifications first though.

> Yes, that's possible: it should produce a WebAssembly.Module as the exported value (symmetric to how you'd get a WebAssembly.{Function,Table,Memory,...} for other export types).

So say one had some complex adapter module chunking workflow where one adapter module acts as a shared chunk exporting multiple module types. I'm just wondering if it would be useful / possible for other adapter modules to be able to in turn import those named exported module types in such a scenario?

> Does the Module Specifier logic have any normalization logic so that you could have two literally-different import strings but they both turn into the same normalized string

The `as "wasm-module"` property is part of the module cache key. So the unique module identifier = id + reflection type combination. Might it make sense to similarly dedupe modules by id + reflection type here? You're exactly right about the sort of workaround in that you can always just redirect through another specifier string, although that might in a performance sensitive loading scenario result in multiple network requests. Again, I'm not really sure there's an interesting use case here, just throwing the edge cases at it.

Reply:
> I'm just wondering if it would be useful / possible for other adapter modules to be able to in turn import those named exported module types in such a scenario?

Ahh, I think I see what you're getting at -- yes, that is a good corner case to investigate.  I think we're good here: if I have some `a.wasm` that exports a nested module named `foo`, then I can write a `b.wasm` that imports `foo` from `a.wasm` via ESM-integration by writing `(import "./a.wasm" (instance (export "foo" (module ...))))`.  In this case, we're not triggering `as "wasm-module"` import case -- rather, we're doing a normal import of the Module Specifier `./a.wasm` and then, given the resulting Module Record, extracting the `foo` field, and expecting a `WebAssembly.Module` to satisfy the module-typed `foo` field (which doesn't involve the ESM loader, we're just looking at the value of the live binding).

Incidentally, this makes me realize a general commutativity relation that should be satisfied by ESM-integration and Module Linking: two adapter modules linked via Module Linking should work the same way as if the same two adapter modules were linking via ESM-integration.  ESM-integration will wrap the exports of the first adapter module with JS API wrappers that get written into the live bindings, and then read by the ESM-loader to provide to the second adapter module which then unwraps and type-checks them -- the point is that this extra intermediate wrap+unwrap step shouldn't make a difference when both sides are wasm (and thus could hypothetically be unobservably optimized away by a smart ESM-loader).  I thiiink we satisfy this property (I can't think of a counterexample), but it's worth thinking about more.

> Might it make sense to similarly dedupe modules by id + reflection type here?

Sorry, could you explain your idea in a bit more detail; I don't quite follow.

Reply:
So in a sense, `(import './foo.wasm') (module $foo))` could possibly be thought of as sugar for `(import './foo.wasm (instance $i (export "default" (module ...)))) (alias $i "default" (module $foo))` plus an extra bit of information that gets sent to the ESM integration to infer the `as "wasm-module"` part.

Perhaps it could be worth supporting the same sugar for a JS module exporting `export default WasmModule`? So effectively just treating this as the general rule for anything not an instance type.

For the other bit of information with regards to the `as "wasm-module"` it might be worth associating each import with a couple of extra bytes. For example if Wasm did want to import CSS or JSON it would want to be able to associate import assertions with imports. Reserving a separate bit space for these kinds of attributes in future might be useful, where initially just normal import and `as "wasm-module"` could be specified.

Reply:
> Perhaps it could be worth supporting the same sugar for a JS module exporting `export default WasmModule`?

That's an interesting idea, but I think it conflicts with the goal of treating JS/wasm modules symmetrically: if I write `(import './foo' (module $foo))`, then ESM will fetch `./foo`, then if the response is `application/wasm`, ESM will definitely just decode+validate and supply the uninstantiated `module`.  But otherwise, does ESM instantiate the JS module to get its `default` export?  If so, then if JS ever in the future added an uninstantiated-parameterized-module-like concept, I wouldn't be able to import it with `module` type (which would otherwise be nice and symmetric), b/c I'm forced to always instantiate the JS module.  It seems safer in the short-term to have an error if a module-typed import isn't given an `application/wasm` response.

> it might be worth associating each import with a couple of extra bytes

I agree that the the extra bit of information needs to be *passed* to the ESM loader, but I think no extra bits of information are necessary in the adapter module itself -- you get the information you need from the import type.  CSS and JSON imports are a useful case to think about, but I think we'd want to use types to express those too.  One option would be to use type imports + built-in modules, so you'd write something like:
```wasm
(adapter module
  (import "std:types:json-module" (type $JSON))
  (import "foo" (global (ref $JSON)))
)
```
Another option would be some standardized custom section interpreted by the ES loader that annotates imports (which is an idea that keeps coming up, but I keep feeling like it's an option of last resort when there's not an alternative rationalized in terms of wasm concepts).

Reply:
This all sounds very sensible to me, thanks for clarifying.